File: herminedb-driver/src/main/java/org/hermine/internal/db/driver/HermineConnectionProperty.java
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol,
 * and explicitly declared author of this file if provided.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.hermine.internal.db.driver;

import jdk.incubator.sql2.ConnectionProperty;
import reactor.core.scheduler.Scheduler;
import reactor.scheduler.forkjoin.ForkJoinPoolScheduler;

import java.util.function.Function;

public enum HermineConnectionProperty implements ConnectionProperty {

    /**
     *
     */
    SCHEDULER(Scheduler.class,
            v -> v instanceof Scheduler,
            ForkJoinPoolScheduler.create("HermineScheduler"),
            false);

    private final Class<?> range;
    private final Function<Object, Boolean> validator;
    private final Object defaultValue;
    private final boolean isSensitive;

    private HermineConnectionProperty(Class<?> range,
                                   Function<Object, Boolean> validator,
                                   Object value,
                                   boolean isSensitive) {
        this.range = range;
        this.validator = validator;
        this.defaultValue = value;
        this.isSensitive = isSensitive;
    }

    @Override
    public Class<?> range() {
        return range;
    }

    @Override
    public boolean validate(Object value) {
        return validator.apply(value);
    }

    @Override
    public Object defaultValue() {
        return defaultValue;
    }

    @Override
    public boolean isSensitive() {
        return isSensitive;
    }
}

File: herminedb-driver/src/main/kotlin/org/hermine/internal/db/driver/HermineConnectionProperty.kt
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol
 * and explicitly declared author of each file if specified.
 * Use of this source code is governed by the Apache 2.0 license.
 */

package org.hermine.internal.db.driver

import jdk.incubator.sql2.ConnectionProperty
import kotlinx.coroutines.experimental.DefaultDispatcher
import java.util.function.Function
import kotlin.coroutines.experimental.CoroutineContext

internal object CONTEXT : HermineConnectionProperty(
        "CONTEXT",
        CoroutineContext::class.java,
        Function { v -> v is CoroutineContext },
        DefaultDispatcher,
        false)

internal abstract class HermineConnectionProperty(
        val name: String,
        val range: Class<*>,
        val validator: Function<Any, Boolean>,
        val defaultValue: Any,
        val sensitive: Boolean) : ConnectionProperty {

    override fun name() = name

    override fun range(): Class<*> {
        return range
    }

    override fun validate(value: Any): Boolean {
        return validator.apply(value)
    }

    override fun defaultValue(): Any {
        return defaultValue
    }

    override fun isSensitive(): Boolean {
        return sensitive
    }
}
